adapter-socket.go

```diff
@@ -0,0 +1,196 @@
+package ipp
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/user"
+	"strconv"
+)
+
+var socketNotFoundError = errors.New("unable to locate CUPS socket")
+var certNotFoundError = errors.New("unable to locate CUPS certificate")
+
+var (
+	DefaultSocketSearchPaths = []string{"/var/run/cupsd", "/var/run/cups/cups.sock", "/run/cups/cups.sock"}
+	DefaultCertSearchPaths   = []string{"/etc/cups/certs/0", "/run/cups/certs/0"}
+)
+
+const defaultRequestRetryLimit = 3
+
+type SocketAdapter struct {
+	host              string
+	useTLS            bool
+	SocketSearchPaths []string
+	CertSearchPaths   []string
+	requestRetryLimit int
+}
+
+func NewSocketAdapter(host string, useTLS bool) *SocketAdapter {
+	return &SocketAdapter{
+		host:              host,
+		useTLS:            useTLS,
+		SocketSearchPaths: DefaultSocketSearchPaths,
+		CertSearchPaths:   DefaultCertSearchPaths,
+		requestRetryLimit: defaultRequestRetryLimit,
+	}
+}
+
+//DoRequest performs the given IPP request to the given URL, returning the IPP response or an error if one occurred
+func (h *SocketAdapter) SendRequest(url string, r *Request, _ io.Writer) (*Response, error) {
+	// set user field
+	user, err := user.Current()
+	if err != nil {
+		return nil, fmt.Errorf("unable to lookup current user: %v", err)
+	}
+	r.OperationAttributes[AttributeRequestingUserName] = user.Username
```

Comment:
the requesting user name should not be set be set in the ipp library

Reply:
i addressed these